middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,101 @@
+import socket
+
+from typing import Iterator
+from typing import Optional
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+from middleman_protocol.utils import append_frame_separator
+from middleman_protocol.utils import escape_decode_raw_message
+from middleman_protocol.utils import escape_encode_raw_message
+
+
+def split_stream(connection: socket.socket) -> Iterator[bytes]:
+    """
+    Lowest level receiver which determines if received data is frame separator,
+    which means previously received data should be yielded,
+    or if received data is something else and receiver should continue to gather bytes.
+    """
+    assert isinstance(connection, socket.socket)
+
+    received_data = b''
+
+    while True:
+        next_bytes = connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        for next_byte in next_bytes:
+            if next_bytes == FRAME_SEPARATOR:
+                yield received_data
+                received_data = b''
+            else:
+                received_data += bytes([next_byte])
+
+
+def unescape_stream(connection: socket.socket) -> Iterator[Optional[bytes]]:
+    """
+    Middle level receiver which determines if received data has broken escaping and yields None to indicate it,
+    or reverses escaping and yields unescaped frame.
+    """
+    assert isinstance(connection, socket.socket)
+
+    for unescaped_data in split_stream(connection=connection):
+        assert isinstance(unescaped_data, bytes)
+
+        # If there is an escape character which is not a valid escape sequence,
+        # yield special value indicating a frame with broken escaping and continue the outer loop.
+        # This can be checked by comparing sum of occurrences of escape character
+        # to sum of occurrences of escape sequences. If they are equal, data is escaped correctly.
```

Comment:
I think this assert can also exist in `constants.py`. I will add it there. :+1: done

Reply:
I think it should be here because there's nothing in `constants.py` that requires it. It's only this specific implementation of the generator that makes this assumption. You could implement it in a different way and this assumption would not be needed. It would be confusing to put it in `constant.py`, out of context.

Reply:
I would say it is specific for the whole concept of library, and as it only checks elements which are defined in constants, it would fit there. I have moved it into generator, you can see how it looks now. For me it looks little weird there, because of the fact that we check constants away from the place in which they are defined. Somebody might easily skip this `assert` when changing things only in `constants.py`.

Reply:
I don't agree that it's specific to the whole concept. This is the only part of the code that makes an assumption that `\\` is not a valid escape character. The protocol does not require it and it would be easy to change the implementation in the future to support this sequence.

Putting it next to the code that makes the assumption does not seem weird to me. Assert should always be next to the code that will break if it's violated. For me putting it next to a constant definition is only a shortcut for cases when the assumption is so general that we would have to keep spamming this assert basically everywhere.

I'm a little surprised because putting assertions next to a value that has just been defined is what always seemed like something someone could call out as weird. Just look at an example:
``` python
SEPARATOR = "abc"
assert len(separator) == 3
```
This looks pointless at a glance because it's obvious that it will never fail. This is still very useful because it serves as a hint that it's 3-characters long for a reason but I think it's less natural and self-explanatory than putting it next to the code that requires it.

> Somebody might easily skip this `assert` when changing things only in `constants.py`.

That's a feature :) It will fail anyway because we're aiming at 100% coverage (if not, our tests suck). But when it's close to the code it's easy to keep it in sync with the code.

Also, there's usually no reason not to put an `assert` in multiple places. If all of them make the same assumption, all should have it.